### ISSUE TITLE
KEDA won't scale when Secret is created by Secrets Store CSI Driver

### ISSUE THREAD

**Issue by kmdnexusvsts** [kedacore/keda#2315]
### Report

The problem is similar to the one described here: https://github.com/kedacore/keda/issues/555 when Deployment depends on Secret object that is not yet created. In our case the Secret is created by Secrets Store CSI Driver (https://secrets-store-csi-driver.sigs.k8s.io/topics/sync-as-kubernetes-secret.html) so it cannot exist before a pod is created. One of the solution suggested in issue 555 was to mark the Secret as optional but in our case this is a risk - we would like to avoid situation when pod is in Running state without the Secret mounted due to i.e. human error. Currently the workaround is to do one-time manual scaling of Deployment replicas and then let KEDA to take over the scaling.

### Expected Behavior

KEDA scales Deployment even if Secret objects not marked as optional are missing. There can be a env variable to keda operator that steers this behavior whether to allow or disallow missing Secret objects.

### Actual Behavior

KEDA won't scale a Deployment if Secret not marked as optional is missing.

$ k get scaledobjects.keda.sh
NAME                              SCALETARGETKIND      SCALETARGETNAME                   MIN   MAX   TRIGGERS   AUTHENTICATION   READY   ACTIVE    AGE
api-server-pool-2                 apps/v1.Deployment   api-server-pool-2                 2     3     cron                        False   Unknown   2m1s

### Steps to Reproduce the Problem

1. Configure Secret creation as described in https://secrets-store-csi-driver.sigs.k8s.io/topics/sync-as-kubernetes-secret.html but just a missing Secret is enough to verify that KEDA won't scale.
2. Configure ScaledObject to scale the Deployment that uses the Secret as envs.

### Logs from KEDA operator

2021-11-22T17:24:51.359Z	ERROR	controller	Reconciler error	{"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledObject", "controller": "scaledobject", "name": "api-server-pool-2", "namespace": "xxxxxxxx", "error": "error resolving secrets for ScaleTarget: error reading secret ref &SecretEnvSource{LocalObjectReference:LocalObjectReference{Name:api-server-pool-2-xxxxxx-envs,},Optional:nil,} on namespace xxxxxxxx: Secret \"api-server-pool-2-xxxxxx-envs\" not found"}

### KEDA Version

2.2.0

### Kubernetes Version

1.21

### Platform

Microsoft Azure

### Scaler Details

Cron

**Comment by zroubalik** [comment]
Related: https://github.com/kedacore/keda/issues/2393

**Comment by asenyaev** [comment]
Hi @zroubalik!

Yes, it's related to https://github.com/kedacore/keda/issues/2393, but there is no solution for Secrets Store CSI Driver usage. 

Usually, a pod for Secrets Store CSI Driver is up when a pod for a deployment has been started and request secret definitions which will be created as secrets and mounted with csi-driver. However, if keda throws an error and cannot scale a deployment, then csi-driver pod is not running too and secrets won't be created. 

The only solution is to keep `minReplicaCount: 1` and in this case csi-driver will be scaled up, secrets will be alive and keda can scale other deployments.

My log of `keda-operator` pod:
```
ERROR   controller.scaledobject Failed to ensure HPA is correctly created for ScaledObject      {"reconciler group": "keda.sh", "reconciler kind": "ScaledObject", "name": "<deployment>", "namespace": "<namespace", "error": "error resolving secrets for ScaleTarget: error resolving secret name &SecretKeySelector{LocalObjectReference:LocalObjectReference{Name:<secret>,},Key:<SECRET_KEY>,Optional:nil,} for env <SECRET_VARIABLE> in namespace <namespace>"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:114
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:311
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/internal/controller/controller.go:227
```

**Comment by zroubalik** [comment]
@asenyaev thanks for the additional information. If this is the way how Secrets Store CSI Driver works we have a chicken & egg problem :) 

So you are right, the only option is to keep `minReplicaCount: 1` or greater.

@tomkerkhove FYI, is there anything we can do on this?

**Comment by tomkerkhove** [comment]
No idea, I have not worked with CSI driver yet - Sorry!

**Comment by tomkerkhove** [comment]
Something we should add ot the FAQ then?

**Comment by zroubalik** [comment]
@tomkerkhove yeah, FAQ or a note somewhere in the Authentication docs makes sense

**Comment by tomkerkhove** [comment]
Are you willing to document this @kmdnexusvsts?

**Comment by mor-benhamo** [comment]
"minReplicaCount: 1" does not work for scaledjob from type "jobTargetRef" as it fails on validation before even creating the first job. do you have any way to overcome it?

**Comment by skibish** [comment]
Facing currently the same challenge that suggestion on `ScaledJob` with `jobTargetRef` to have `minReplicaCount: 1` is not working. Job is not willing to start

**Comment by waterbuckit** [comment]
I have also made attempts to get this work by having secret be created as part of another deployment. The problem is that it still isn't there in time for KEDA's checks to see if it can create the job. KEDA tries 10 times in rapid succession, but then waits 10 minutes (by which time the secret is created).

Is there any way to configure this wait time?

**Comment by tomkerkhove** [comment]
We're happy to fix this but spread thin. Anyone who wants to propose a plan and contribute the improvement?
